### PR TITLE
feat: mobile auth

### PR DESCRIPTION
**Pull request by Nirajn2311**
The new endpoint will generate loopback accessTokens by using the given auth0 access tokens.

### CODE REVIEW

File: api-server/src/common/models/user.js
Comment (ojeytonwilliams) [inline api-server/src/common/models/user.js:349]:
```suggestion
      this.createAccessToken(77760000000, (err, accessToken) => {
```
So, a couple of things.

We don't necessarily need to set this here.  We can modify `User.maxTTL` in https://github.com/freeCodeCamp/freeCodeCamp/blob/1155e8f2de435a3f43c6086e62a855226e15bea4/api-server/src/common/models/user.js#L164 

These should be _seconds_ not milliseconds.

cc @raisedadead

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: api-server/src/common/models/user.js
Comment (raisedadead) [reply api-server/src/common/models/user.js:349]: I am unsure if the `User.settings.ttl` even works. I would not rely on it.

**Comment by Nirajn2311** [comment]
Rate limit is set to a maximum of 10 requests per IP per 15 minute window

File: api-server/src/server/boot/authentication.js
Comment (raisedadead) [inline api-server/src/server/boot/authentication.js:7]: Can't use this because:

> This module does not share the state with other processes/servers by default.

**Review by raisedadead — CHANGES_REQUESTED** [review]

File: api-server/src/server/boot/authentication.js
Comment (Nirajn2311) [reply api-server/src/server/boot/authentication.js:7]: What if it used a MongoDB store just like the session middleware? Will it work in that case?

Comment (raisedadead) [reply api-server/src/server/boot/authentication.js:7]: Yes - probably. I also realized that we might not be collecting end-user IP addresses.

Comment (ojeytonwilliams) [reply api-server/src/server/boot/authentication.js:7]: In principle it could work without sharing state, though it could be frustrating if a user is unlucky and keeps hitting the same machine.  That's rather unlikely, though.

Comment (ojeytonwilliams) [reply api-server/src/server/boot/authentication.js:7]: Well, yes, obviously. My point was just that that doesn't require synchronisation.

If you want a max 10 requests in 15 minutes globally you can either synchronize or allow 10/n per instance.

I'm not saying this is perfect, mind, since the same user could get routed the same VM repeatedly. I doubt many users would run into this, though.

Comment (raisedadead) [reply api-server/src/server/boot/authentication.js:7]: 10 requests / 15 min from a given unique individual user (maybe email / or IP as an identifier) regardless of the instance they reach.

Comment (ojeytonwilliams) [reply api-server/src/server/boot/authentication.js:7]: If we have 2 instances which each allow 5 requests per 15 minutes from a unique user then that's 10 max.  Since it's random then, without state sharing, you could start rejecting requests after 5 so it's not identical, but the max is still 10.

Obviously there are tradeoffs (you have to adjust as you increase/decrease instances), but I'm not seeing an in principle reason why this impossible.

Comment (raisedadead) [reply api-server/src/server/boot/authentication.js:7]: > but I'm not seeing an in principle reason

It's over-engineering because now you are guessing the rate based on some fancy way to get instance count (or involve a human to set some parameter etc.).

Anyways we have entered bikeshedding territory here.

The approach that we want now is:

1. Introduce rate-limiting via simple state management or some other alternative (like debouncing, etc.) 
2. The lib in this current change claims explicitly to use others instead of them.

Comment (Nirajn2311) [reply api-server/src/server/boot/authentication.js:7]: > I also realized that we might not be collecting end-user IP addresses.

Won't the IP be in the request object?

> The lib in this current change claims explicitly to use others instead of them.

They say that they don't share state by default ie. in-memory store

Comment (raisedadead) [reply api-server/src/server/boot/authentication.js:7]: > They say that they don't share state by default ie. in-memory store

Cool. Let's try this as is for now and make a call later if we need to switch.

Comment (Nirajn2311) [reply api-server/src/server/boot/authentication.js]: > Cool. Let's try this as is for now and make a call later if we need to switch.

I already implemented it 😅 

> We are not forwarding the end-user IP to the API

Is there any issue in forwarding the user IP if we go about with it?

**Comment by Nirajn2311** [comment]
I've integrated the rate limit into Loopbacks middleware flow so we can easily use it with any other route if required.

File: api-server/src/server/boot/authentication.js
Comment (raisedadead) [reply api-server/src/server/boot/authentication.js]: It should be possible in theory, and I remember doing it for the forum years ago. I need to go back in time and trace the correct headers from Cloudflare and forward them to the API.

**Comment by Nirajn2311** [comment]
It won't interfere with any other routes. It's set for only the `/mobile-login` route.

File: api-server/src/common/models/user.js
Comment (raisedadead) [reply api-server/src/common/models/user.js]: I would not rely on it, pretty sure it's not used.

File: api-server/src/server/boot/authentication.js
Comment (raisedadead) [inline api-server/src/server/boot/authentication.js:207]: Axios is a particularly heavy binary. I do not see it being used other than this call. This could be done in vanilla JavaScript instead.

File: api-server/src/common/models/user.js
Comment (raisedadead) [inline api-server/src/common/models/user.js:166]: This has no effect. You should not depend on it, and remove this instead.

File: api-server/src/server/boot/authentication.js
Comment (raisedadead) [inline api-server/src/server/boot/authentication.js:90]:
```suggestion
  api.get('/mobile-login', ifNotMobileRedirect, ifUserRedirect, mobileLogin(app));
```

You should create a handler called `ifNotMobileRedirect` that checks for the request from mobile devices (non-desktop browsers) and then reply with non-availability.

It could be user-agent-based for now.

**Review by raisedadead — CHANGES_REQUESTED** [review]

File: api-server/src/server/boot/authentication.js
Comment (Sembauke) [reply api-server/src/server/boot/authentication.js:207]: When I tried using fetch instead of Axios it said that fetch was not defined. Normally you use Node fetch or Axios to do this.

Comment (raisedadead) [reply api-server/src/server/boot/authentication.js:207]: Correct. 

It would be best if you had https://www.npmjs.com/package/node-fetch (Node.js v16 and below), which is much lighter than Axios. 

There is also a native fetch API in Node.js v18 (we are moving to this very shortly). See: https://nodejs.org/de/blog/announcements/v18-release-announce

File: api-server/src/common/models/user.js
Comment (ojeytonwilliams) [reply api-server/src/common/models/user.js:166]: It did seem to matter when I tested the new endpoint.  Without adding this I got back a shorter TTL.

File: api-server/src/server/boot/authentication.js
Comment (Sembauke) [reply api-server/src/server/boot/authentication.js:207]: Yes, Axios was already a dependency.

Comment (ojeytonwilliams) [reply api-server/src/server/boot/authentication.js:207]: Yep. Thinking about it, there are two paths forwards that seem fine to me.

1. use node-fetch here and migrate donation to node-fetch separately
2. use axios here and migrate both to node-fetch in another PR

Maybe 1. is best because it limits the scope. Also, the api already has `node-fetch` as a transitive dependency so having it explicitly will not change anything about the installation.

File: api-server/src/common/models/user.js
Comment (Nirajn2311) [reply api-server/src/common/models/user.js:166]: Also the cookie lifetime was shortened without this line

File: api-server/src/server/boot/authentication.js
Comment (Nirajn2311) [reply api-server/src/server/boot/authentication.js:211]: I'm in favour of 1 as it's better to use node-fetch now itself and well not forget to change from Axios to node-fetch in future

File: api-server/src/common/models/user.js
Comment (ojeytonwilliams) [reply api-server/src/common/models/user.js:166]: (I unresolved this to make sure Mrugesh has a chance to see the discussion)

**Comment by raisedadead** [comment]
We are officially on Node.js 18 now. We have changed the merge requirements accordingly.

You will need to rebase this PR against the main branch OR Click the "Update branch" button to get the new checks going.

Thanks for your contribution.

File: api-server/src/server/utils/middleware.js
Comment (raisedadead) [inline api-server/src/server/utils/middleware.js:87]:
```suggestion
    //
    // Todo: Use the below check once we have done more research on usage
    //
    // const isMobile = /(iPhone|iPad|Android)/.test(req.headers['user-agent']);
    // if (!isMobile) {
    //  res.json({ error: 'not from mobile' });
    // } else {
    //  next();
    // }
    next();
```

I am having second thoughts about using this because, on the surface, it looks very brittle. There are several other devices that we would need to account for. 

Also, regex is a bad idea in principle as the device list grows.

Let's shelf this for now and revisit it in the future.

**Review by raisedadead — CHANGES_REQUESTED** [review]

File: api-server/src/server/utils/middleware.js
Comment (Nirajn2311) [reply api-server/src/server/utils/middleware.js:87]: Cool that sounds good to me 👍

**Review by raisedadead — APPROVED** [review]
Looks good to me. I want a review from the rest of the folks on @freeCodeCamp/dev-team who have been requested a review.

We will probably end up doing a gated deployment for this PR.

File: api-server/src/server/boot/authentication.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/authentication.js:221]: What's the idea behind this redirect? Mobile users can't (or shouldn't?) be trying to use the web signin, so why redirect there?

File: api-server/src/server/middlewares/request-authorization.js
Comment (ojeytonwilliams) [inline api-server/src/server/middlewares/request-authorization.js:29]:
```suggestion
const mobileLoginRE = /^\/mobile-login\/?$/;
```
I think we can be relaxed about the trailing slash, but should be stricter otherwise.

File: package.json
Comment (ojeytonwilliams) [inline package.json:118]: invariant belongs here, but the two new dependencies should be in api-server.

File: api-server/src/server/boot/authentication.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/authentication.js:229]:
```suggestion
        await user.mobileLoginByRequest(req, res);
        res.end();
```
Since the token is getting added as a cookie I don't believe there's any need to send it back in the body.

**Review by ojeytonwilliams — DISMISSED** [review]
This is looking good. I didn't spot anything dangerous, but there were a few minor issues:

File: api-server/src/server/boot/authentication.js
Comment (Nirajn2311) [reply api-server/src/server/boot/authentication.js:221]: Good catch. There won't be a redirect, we'll be directly stopping the flow there.

**Comment by Nirajn2311** [comment]
I had missed out on 2 edge cases and those have been fixed now. The edge cases were:
* Invalid Auth0 token is sent in the request which in turn leads to a 401 response from Auth0
* A new user attempts to login from the mobile which in turn lead to "user not found" error.

File: api-server/src/server/boot/authentication.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/authentication.js:204]:
```suggestion
      headers: { token }
```
Nit: no need to reassign

File: api-server/src/server/boot/authentication.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/authentication.js:210]:
```suggestion
        headers: { Authorization: `Bearer ${token}` }
```

File: api-server/src/server/boot/authentication.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/authentication.js:239]:
```suggestion
    return User.findOne$({ where: { email } })
      .do(async user => {
        if (!user) {
          user = await User.create({ email });
```

**Review by ojeytonwilliams — APPROVED** [review]
Looks good, guys.  All I found were stylistic issues, i.e. nothing we need to block this for.

**Comment by Nirajn2311** [comment]
Yes, I'll be working on the new user flow in a separate PR at the mobile side.

**Review by Sembauke — APPROVED** [review]

**Comment by Sembauke** [comment]
@ojeytonwilliams , @raisedadead can we merge this?

**Comment by ojeytonwilliams** [comment]
@Sembauke I think the code looks good. An extra pair of eyes wouldn't hurt, though. @moT01 could you give this a quick look?

**Review by moT01 — APPROVED** [review]
LGTM 🎉 

We want to wait to merge this, as @raisedadead mentioned, so we can do a dedicated deployment with only these changes I think. So maybe we should probably wait for him.

File: api-server/src/server/boot/authentication.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/authentication.js:202]: Since this is async, express's error handling can't work properly (not express 4's, anyway!).  Could you wrap the whole thing in a try/catch block and call next(err)?

File: api-server/src/server/boot/authentication.js
Comment (ojeytonwilliams) [inline api-server/src/server/boot/authentication.js:236]:
```suggestion
    User.findOne$({ where: { email } })
```
This implies it meaningfully returns something, but it doesn't.  Express middlewares only interact via res, req or next - the return value is ignored.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Just some minor things @Nirajn2311. We should be ready to merge after you've addressed them.

File: api-server/src/server/boot/authentication.js
Comment (Nirajn2311) [reply api-server/src/server/boot/authentication.js:206]: This is not a bearer token and is the opaque token we get after the user logs in. That is then used to exchange for FCCs access token and immediately deleted from the user mobile once the whole process is done.

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1:

**Review by raisedadead — CHANGES_REQUESTED** [review]
Hold for dedicated deployment. May need creating collection ahead of time.

**Pull request merged**
